### ISSUE TITLE
Rspec under rails 4.1.8 doesn't apply rollback transaction to secondary database

### ISSUE THREAD

**Issue by jeremywadsack** [rails/rails#17776]
Our app has two databases, the second of which is shared with another app. My `database.yml` looks like this for the test environments:

```
test:
  adapter: postgresql
  database: my-app-test
  host: localhost
  encoding: utf8
  min_messages: WARNING

shared_test:
  adapter: postgresql
  database: shared-test
  host: localhost
  encoding: utf8
  min_messages: WARNING
```

Models that use the shared database inherit from this:

```
class ServiceModel < ActiveRecord::Base
  self.abstract_class = true
  establish_connection :"shared_#{Rails.env}"

  def readonly?
    true if persisted?
  end

  def delete
    raise ActiveRecord::ReadOnlyRecord
  end

  def destroy
    raise ActiveRecord::ReadOnlyRecord
  end
end
```

We run rspec with transactional fixtures but the rspec team states that this is [a wrapper around Rails configuration](https://github.com/rspec/rspec-rails/issues/1235#issuecomment-64489255):

```
  config.use_transactional_fixtures = true
```

When I run `rspec` under rails 3.2.19, everything runs fine and I can run the tests over and over and they continue to pass.

When I run `rspec` under rails 4.1.8, they pass the first time, but running them again raises lots of `PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint` errors because the shared test database was not cleared out.

I have inspected the logs from postgresql for both branches and have made [gists](https://gist.github.com/jeremywadsack/8f3e95f4e26194686d7e) available that show that ROLLBACK is being called for the shared database under rails 3 but not under rails 4. Specifically, [line 235 in the rails_3.2.19](https://gist.github.com/jeremywadsack/8f3e95f4e26194686d7e#file-rails_3-2-19-L235) it calls ROLLBACK on the shared database (there's a corresponding BEGIN at line 98). This is missing in the rails 4.1.8 version.

**Comment by magikid** [comment]
@jeremywadsack I'd say send in a pull request if you're still having this issue. Did you ever resolve it?

**Comment by jeremywadsack** [comment]
@magikid  We just enabled `eager_load` in test as that was the "fastest" way to resolve this.
I missed the link from 17921; although I don't understand enough of that to know whether fixing this would address that at all. 

Are you having that problem? I could work on a PR, but if we're the only people with two databases, using rspec, that want to no eager load tests (given the lack on response on this and the age of the code change in Rails 4), then I don't want to do extra work that no one will need.

**Comment by magikid** [comment]
I'm not having this problem. I'm just working through stale issues.

**Comment by jeremywadsack** [comment]
I appreciate that. I'll take a stab at a fix through instrumentation.

**Comment by jeremywadsack** [comment]
@magikid Pull Request opened to fix this but Travis doesn't want to re-try the build. Do you know what I need to do to kick it?

**Comment by jeremywadsack** [comment]
I think matthewd fixes the travis issue by closing and re-opening the PR. Travis has run and passed all tests, so that PR is ready for review/test/merge.

**Comment by greggawatt** [comment]
The PR that solve this doesn't seem to have gotten merged in yet and currently has conflicts. @jeremywadsack are you planning to take another stab at it?

**Comment by jeremywadsack** [comment]
@greggawatt: Sure, I can rebase on current master. (I'm pretty sure there were no merge conflicts when I submitted on Jul 20).

**Comment by jeremywadsack** [comment]
PR squashed and rebased on current master.

I have also updated the sample application at https://github.com/jeremywadsack/rollback-test with a `rails_5_beta` branch that proves the fix.

**Comment by jeremywadsack** [comment]
What's the expectation for this? I have now updated and rebased #20818 twice per your requests but no one has merged it yet. It is now in a conflict state again. I can rebase again but I'm hesitant to do that unless there's someone with time to merge it.

**Comment by rafaelfranca** [comment]
@jeremywadsack Sorry, the bot is automatic but also requires a manual intervention to not mark issues as stale. I added the attached PR label so it will not mark it stale again.

**Comment by jeremywadsack** [comment]
@rafaelfranca thanks for the quick response. I'd love to see this issue resloved. Anything I can do to help make that easier?

**Comment by eugeneius** [comment]
Can this be closed, since https://github.com/rails/rails/pull/20818 was merged?

Nice work by the way @jeremywadsack! 🙏

**Comment by maclover7** [comment]
👍

**Comment by njt1982** [comment]
Is this only fixed in Rails 5? I'm on `4.2.5.2` and have this very problem...

**Comment by jeremywadsack** [comment]
@njt1982 Based on https://github.com/rails/rails/pull/24954#issuecomment-233831517 I do not believe that this is ported back to 4.2.
